### PR TITLE
(PROJ): use single Rust module type

### PR DESCRIPTION
**Pull request by matklad**
A single `Cargo.toml` can specify 0-1 library crates and 0-n binary crates, so it doesn't make much sense to have distinct library and executable modules.

I use the following terminology here:

`mod` --- rust module like in `pub mod foo { }`.

`module` --- an IntelliJ object, representing a bunch of rust sources controlled by a single `Cargo.toml`.

`crateRoot` --- a top level `mod`, from which the compilation starts. It is also called `target` in `cargo metadata`.

A user visible effect of this change is that crates like [crossbeam](https://github.com/aturon/crossbeam) which contain both `bin` and `lib` tragets now resolve better, because lib crate is not shadowed by bin crate.

@alexeykudinkin please review

### CODE REVIEW

**Comment by matklad** [comment]
Fixed the tests. I always trip over eager initialization :(

**Comment by matklad** [comment]
ping

File: src/main/kotlin/org/rust/lang/core/psi/util/RustModExtensions.kt [hunk 1]
```diff
@@ -77,13 +79,7 @@ internal val RustModItem.ownedDirectory: PsiDirectory?
...
 
 private val RustModItem.isCrateRoot: Boolean
-    get() {
-        val file = containingFile.virtualFile ?: return false
-        val containingDir = file.parent
-        return project.getCrateSourceRootFor(file)?.let {
-            it.equals(containingDir) && (containingFile.name == RustModules.MAIN_RS || containingFile.name == RustModules.LIB_RS)
-        } ?: false
-    }
+    get() = getModule()?.crateRoots.orEmpty().contains(this)
```
Comment (alexeykudinkin) [inline src/main/kotlin/org/rust/lang/core/psi/util/RustModExtensions.kt]: Bringing up PSI for all the `main.rs`/`lib.rs` files just to check whether this particular `mod` is the one of them seems fishy.

Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/psi/util/RustModExtensions.kt]: yeah, fixed.

File: src/main/kotlin/org/rust/cargo/project/module/util/RustCrateUtil.kt [hunk 2]
```diff
@@ -17,23 +14,17 @@ object RustCrateUtil
...
     ModuleRootManager.getInstance(this).getSourceRoots(includingTestRoots).toList()
 
-val Module.rootMod: RustModItem?
-    get() = getSourceRoots(includingTestRoots = false).firstOrNull()?.let {
-        var target = when (getOptionValue(Module.ELEMENT_TYPE)) {
-            RustExecutableModuleType.MODULE_TYPE_ID -> RustModules.MAIN_RS
-            RustLibraryModuleType.MODULE_TYPE_ID    -> RustModules.LIB_RS
-            else                                    -> return null
-        }
+val Module.crateRootFiles: Collection<VirtualFile>
```
Comment (alexeykudinkin) [inline src/main/kotlin/org/rust/cargo/project/module/util/RustCrateUtil.kt]: This doesn't seem as a very handy method to be exposed as an API. I'd rather replace extension method with an ordinary one and rather added `isCrateRoot(f: VirtualFile)` method.

Comment (matklad) [inline src/main/kotlin/org/rust/cargo/project/module/util/RustCrateUtil.kt]: Fixed,  although `isCrateRoot(f: VirtualFile)` does not seem much handier, it is at least more focused :)

**Pull request merged**